### ISSUE TITLE
[naming-convention] don't quit on first match

### ISSUE THREAD

**Issue by LakMoore** [typescript-eslint/typescript-eslint#2131]
```JS
{
  "rules": {
    "@typescript-eslint/naming-convention": [
          "error",
          {
            selector: "default",
            format: ["camelCase"],
            leadingUnderscore: "forbid",
            trailingUnderscore: "forbid",
          },
          {
            selector: "default",
            modifiers: ["private"],
            format: ["camelCase"],
            leadingUnderscore: "require",
          },
          {
            selector: "property",
            types: ["boolean"],
            format: ["PascalCase"],
            prefix: ['is', 'should', 'has', 'can', 'did', 'will'],
            leadingUnderscore: "allow",
          },
      ],
  }
}
```

```TS
  private isProperty: boolean;
  private hasProperty: string;
```

**Expected Result**
Two lint errors; one for each property declaration.

**Actual Result**
A single lint error on hasProperty

**Additional Info**
Neither property has a leading underscore but both are private.
The rules should be ordered and then tested in order until the first FAILURE is found.  I believe the test is quit after the first matching PASS is found.

Did eslint throw an exception?
No exception thrown
Debug log: https://gist.github.com/LakMoore/2010d94ba845bdcc999f6f2e7eb4fdf2

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.0.2` |
| `@typescript-eslint/parser`        | `3.0.2` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `7.1.0` |
| `node`                             | `12.16.3` |
| `npm`                              | `6.14.4` |

**Comment by bradzacher** [comment]
This is done on purpose, otherwise you could get into cases like this:

```json
[
  {
    selector: "default",
    format: ["camelCase"],
    leadingUnderscore: "forbid",
  },
  {
    selector: "default",
    format: ["camelCase"],
    modifiers: ["private"],
    leadingUnderscore: "require",
  },
]
```
```ts
class foo {
  private _property_ = 1;
}
```

If the rule continues checking after it matches successfully, then this code would error:
- The rule would first check the `"private"` config, and find that the property matches completely as it is camelcase with a leading underscore.
- The rule would then check the other config, and find that the property does not match, and would error because it has a leading underscore.

**Comment by LakMoore** [comment]
Thanks for the feedback @bradzacher.  So how do I arrange my rules so that it is linted correctly?

Surely, the fact that the property is of boolean type should not preclude the fact that it might have a private modifier?

Do I have to create an additional rule for private booleans?

I feel I may not be the last person to raise this issue or expect the rules to perform differently.

**Comment by bradzacher** [comment]
This functionality is explicitly called out in the docs:

https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/naming-convention.md#selector-options

> The ordering of selectors does not matter. The implementation will automatically sort the selectors to ensure they match from most-specific to least specific. It will keep checking selectors in that order until it finds one that matches the name.

---

> Do I have to create an additional rule for private booleans?

Yes. If you want to have a naming convention for a more specific selector, you must define it for that selector.

We talked about adding an inheritance config, but decided to keep the initial options as minimal as possible, as it was already a very complex rule, and we thought reasoning about inheritance would over complicate it. We could potentially add something like that down the line.
